Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_LOCAL_MEMORY_REMOTE_DOCUMENT_CACHE_H_
+#define FIRESTORE_CORE_SRC_LOCAL_MEMORY_REMOTE_DOCUMENT_CACHE_H_
+
+#if !defined(__OBJC__)
+#error "For now, this file must only be included by ObjC source files."
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+
+
+@class FSTLocalSerializer;
+@class FSTMaybeDocument;
+@class FSTMemoryLRUReferenceDelegate;
+@class FSTQuery;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class MemoryRemoteDocumentCache {
+ public:
+  void AddEntry(FSTMaybeDocument *document);
+
+  void RemoveEntry(const model::DocumentKey &key);
+
+  FSTMaybeDocument *_Nullable Get(const model::DocumentKey &key);
+
+  model::MaybeDocumentMap GetAll(const model::DocumentKeySet &keys);
+
+  model::DocumentMap GetMatchingDocuments(FSTQuery *query);
+
+  std::vector<model::DocumentKey> RemoveOrphanedDocuments(
+          FSTMemoryLRUReferenceDelegate *reference_delegate, model::ListenSequenceNumber upper_bound);
+
+  size_t CalculateByteSize(FSTLocalSerializer *serializer);
```

Comment:
Nit: this function can be `const`.

Reply:
As discussed, there will eventually be a leveldb version that will access non-const methods.